### PR TITLE
fix(Scripts/AhnQiraj): Ossirian the Unscarred - Ossirian Crystals imp…

### LINKED ISSUE #12472: [AQ20] Ossirian the Unscarred - Ossirian Crystals spawn and respawn mechanics
### Current Behaviour

- Only one spawns at a time
- The starting crystal is missing.
- The crystals do not despawn after a wipe

### Expected Blizzlike Behaviour

- 6 normal ones are up at all times during the fight, excluding the starting one, which is different
- A special crystal is always spawned before the pull but it **does not** respawn during the fight, and **only** respawns 30s **after a reset** (or when the boss reappears). It is separate from all other crystals and does not take part of the spawning mechanic.
- **All** crystals perform their despawn animation and despawn after a wipe.

Respawn mechanic:
- When a crystal is activated, a random one that is not already up spawns when the former casts its Weakness spell (as soon as the despawn animation starts, which is when the spire goes back down into the sand), so that there is always 6 crystals up at any time.

### Source

Retail, Sniffs.

>there are few of them in the area (5+)
>dont click on it before the boss is there... else it will be "consumed" and it wont respawn ([source](https://www.wowhead.com/object=180619/ossirian-crystal#comments:id=1374817))

>As of a few patches ago, there will always be a first crystal available right in front of Ossirian. This is much easier than the old way, which required you to find a crystal from the get-go. ([source](https://www.wowhead.com/npc=15339/ossirian-the-unscarred#comments:id=608))

>All you have to do is put two druids in a group by themselves and have them run (travel form) in opposite directions tagging crystals that are along the outside and towards the back where the worst tornados are. This forces crystals to pop up in the middle all the time ([source](https://www.wowhead.com/npc=15339/ossirian-the-unscarred#comments:id=8700)) (this one says that activating crystals will respawn others that were already used)

### Steps to reproduce the problem

.go xyz -9397 1947 87 509 2.5

see if the starting crystal is already spawned
activate it, see that it does not respawn until a reset

engage the boss
see if 6 crystals besides the starting one spawn

`.aura 40733` on self to avoid the Summon Player spell
engage the boss
go around activating crystals, see if another spawns after you activate one

### Extra Notes

Total of 11 standard spawns and 1 special spawn (the starting crystal)

Starting Crystal:

```
[3] UpdateType: CreateObject1
[3] Object Guid: Full: 0x2C3B083FA0B062C00048B4000059CB60 GameObject/0 R3778/S18612 Map: 509 Entry: 180619 Low: 5884768
[3] Object Type: 8 (GameObject)
[3] NoBirthAnim: False
[3] EnablePortals: False
[3] PlayHoverAnim: False
[3] HasMovementUpdate: False
[3] HasMovementTransport: False
[3] Stationary: True
[3] HasCombatVictim: False
[3] HasServerTime: False
[3] HasVehicleCreate: False
[3] HasAnimKitCreate: False
[3] HasRotation: True
[3] HasAreaTrigger: False
[3] HasGameObject: False
[3] HasSmoothPhasing: False
[3] ThisIsYou: False
[3] SceneObjCreate: False
[3] HasPlayerCreateData: False
[3] HasConversation: False
[3] PauseTimesCount: 0
[3] Stationary Position: X: -9407.718 Y: 1960.2114 Z: 85.63906
[3] Stationary Orientation: 1.1170093
[3] GameObject Rotation: X: 0 Y: 0 Z: 0.5299187 W: 0.84804845
[3] FieldFlags: 0 (None)
[3] EntryID: 180619
[3] DynamicFlags: 4294901760
[3] Scale: 3
[3] DisplayID: 6486
[3] SpellVisualID: 0
[3] StateSpellVisualID: 0
[3] SpawnTrackingStateAnimID: 1716
[3] SpawnTrackingStateAnimKitID: 0
[3] StateWorldEffectsQuestObjectiveID: 0
[3] CreatedBy: Full: 0x0
[3] GuildGUID: Full: 0x0
[3] Flags: 0
[3] ParentRotation: X: 0 Y: 0 Z: 0 W: 1
[3] FactionTemplate: 35
[3] State: 1
[3] TypeID: 10
[3] PercentHealth: 255
[3] ArtKit: 0
[3] CustomParam: 0
[3] Level: 0
[3] AnimGroupInstance: 0
```

Others:

```
[12] Stationary Position: X: -9432.3955 Y: 1782.5293 Z: 85.63908
[12] Stationary Orientation: 5.864307
[12] GameObject Rotation: X: 0 Y: 0 Z: -0.20791149 W: 0.9781476
```
```
[2] Stationary Position: X: -9299.727 Y: 1748.4532 Z: 85.63907
[2] Stationary Orientation: 1.448622
[2] GameObject Rotation: X: 0 Y: 0 Z: 0.6626196 W: 0.7489561
```
```
[8] Stationary Position: X: -9406.096 Y: 1862.3763 Z: 85.63908
[8] Stationary Orientation: 6.2308264
[8] GameObject Rotation: X: 0 Y: 0 Z: -0.026176453 W: 0.99965733
```
```
[16] Stationary Position: X: -9248.413 Y: 1974.8303 Z: 85.639084
[16] Stationary Orientation: 5.8992143
[16] GameObject Rotation: X: 0 Y: 0 Z: -0.1908083 W: 0.9816273
```
```
[17] Stationary Position: X: -9506.192 Y: 1865.572 Z: 85.63909
[17] Stationary Orientation: 4.2760572
[17] GameObject Rotation: X: 0 Y: 0 Z: -0.8433914 W: 0.53729963
```
```
[21] Stationary Position: X: -9383.292 Y: 2012.6808 Z: 85.65106
[21] Stationary Orientation: 2.932139
[21] GameObject Rotation: X: 0 Y: 0 Z: 0.99452114 W: 0.10453553
```
```
[2] Stationary Position: X: -9282.08 Y: 1887.3401 Z: 85.63908
[2] Stationary Orientation: 2.0071285
[2] GameObject Rotation: X: 0 Y: 0 Z: 0.8433914 W: 0.53729963
```
```
[1] Stationary Position: X: -9244.413 Y: 1808.983 Z: 85.639084
[1] Stationary Orientation: 5.637414
[1] GameObject Rotation: X: 0 Y: 0 Z: -0.3173046 W: 0.94832367
```
```
[2] Stationary Position: X: -9357.861 Y: 1929.0792 Z: 85.63907
[2] Stationary Orientation: 1.0646496
[2] GameObject Rotation: X: 0 Y: 0 Z: 0.50753784 W: 0.8616294
```
```
[4] Stationary Position: X: -9367.171 Y: 1780.8878 Z: 85.639084
[4] Stationary Orientation: 1.902409
[4] GameObject Rotation: X: 0 Y: 0 Z: 0.8141155 W: 0.58070296
```
```
[5] Stationary Position: X: -9188.444 Y: 1940.2103 Z: 85.639046
[5] Stationary Orientation: 3.1765332
[5] GameObject Rotation: X: 0 Y: 0 Z: -0.9998474 W: 0.017469281
```

Despawn & Respawn:

```
ClientToServer: CMSG_GAME_OBJ_USE (0x34FC) Length: 15 ConnIdx: 1 Time: 07/29/2022 19:12:53.792 Number: 686
GameObjectGUID: Full: 0x2C42183FA0B062C0000A980000645ABE GameObject/0 R4230/S2712 Map: 509 Entry: 180619 Low: 6576830
```

```
ServerToClient: SMSG_GAME_OBJECT_DESPAWN (0x25C1) Length: 15 ConnIdx: 1 Time: 07/29/2022 19:12:59.984 Number: 871
ObjectGUID: Full: 0x2C42183FA0B062C0000A980000645ABE GameObject/0 R4230/S2712 Map: 509 Entry: 180619 Low: 6576830
```

```
ServerToClient: SMSG_UPDATE_OBJECT (0x27BA) Length: 32 ConnIdx: 1 Time: 07/29/2022 19:13:00.055 Number: 874
NumObjUpdates: 0
MapID: 509 (509)
HasRemovedObjects: True
DestroyedObjCount: 1
RemovedObjCount: 1
(Destroyed) [0] ObjectGUID: Full: 0x2C42183FA0B062C0000A980000645ABE GameObject/0 R4230/S2712 Map: 509 Entry: 180619 Low: 6576830
Data size: 0
```
```
ServerToClient: SMSG_UPDATE_OBJECT (0x27BA) Length: 796 ConnIdx: 1 Time: 07/29/2022 19:13:00.467 Number: 891
NumObjUpdates: 2
MapID: 509 (509)
HasRemovedObjects: False
Data size: 785
[1] UpdateType: CreateObject1
[1] Object Guid: Full: 0x2C42183FA0B062C0000A980000645B74 GameObject/0 R4230/S2712 Map: 509 Entry: 180619 Low: 6577012
```

### AC rev. hash/commit

https://github.com/azerothcore/azerothcore-wotlk/commit/757375f982429c60b17e9b1f559ed86cb22abbc4

### Operating system

Windows 10

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by UltraNix**
…rovements:

Added missing initial spawn of crystal.
Added more crystal spawn points.
Crystals should despawn on wipe.
Fixes #12472

## Issues Addressed:

- Closes #12472

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

More info [here](#12472)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Maelthyrr** [comment]
Tested:

- [X] All spawn points see
- [X] Crystal despawn when reset/die

I went to tested with: https://github.com/azerothcore/azerothcore-wotlk/pull/12647/files but I got a merge conflict.

**Review by IntelligentQuantum — APPROVED** [review]

**Comment by Gultask** [comment]
- [ ] They only respawn if the last crystal affects the boss with the weakness spell.
- [ ] The starting crystal is respawning immediately but it's not interactable until a reset.

I understand that the number of crystals is a matter for debate.
https://youtu.be/4ekKACUEMAE?t=148 shows 3, one behind a pillar, one to the left and another to the right which they didn't use.
So we should use at least 3 concurrent spawns until we can confirm the correct number in WotLK.

**Comment by UltraNix** [comment]
@Gultask 
`They only respawn if the last crystal affects the boss with the weakness spell.` I don't understand that

**Comment by Gultask** [comment]
If the boss is far away from the crystal once it's been activated. It won't spawn another one.
For me it's only spawning another one if the boss is close to the crystal that's being activated, and thus is affected by the Weakness spells that are cast by the Crystal Triggers

**Comment by UltraNix** [comment]
`The starting crystal is respawning immediately but it's not interactable until a reset.` Show how it should look like

**Comment by UltraNix** [comment]
@Gultask I just tested that and spawned another crystal everytime, just needed to find it.

**Comment by Gultask** [comment]
>Show how it should look like

Basically it should despawn entirely and only respawn after the boss reappears after a reset. Nothing else respawns the starting crystal.

>I just tested that and spawned another crystal everytime, just needed to find it.

Sorry then something is wrong with my AC, I'll clean my DB and try again.

**Comment by Maelthyrr** [comment]
@UltraNix Merge conflict

**Comment by Nefertumm** [comment]
fixed

**Comment by UltraNix** [comment]
@Gultask Please, test it again.

File: src/server/scripts/Kalimdor/RuinsOfAhnQiraj/boss_ossirian.cpp [hunk 1]
```diff
@@ -94,9 +98,33 @@ struct boss_ossirian : public BossAI
...
+
+        Creature* trigger = me->GetMap()->GetCreature(_triggerGUID[0]);
+        if (trigger)
+        {
+            trigger->DespawnOrUnsummon();
+            if (GameObject* crystal = me->GetMap()->GetGameObject(_crystalGUID[0]))
+                crystal->Delete();
+        }
+
+        if (trigger = me->GetMap()->SummonCreature(NPC_OSSIRIAN_TRIGGER, initialCrystalPosition))
```
Comment (acidmanifesto) [inline src/server/scripts/Kalimdor/RuinsOfAhnQiraj/boss_ossirian.cpp:110]: image
```suggestion
        if (trigger == me->GetMap()->SummonCreature(NPC_OSSIRIAN_TRIGGER, initialCrystalPosition))
```

**Review by acidmanifesto — CHANGES_REQUESTED** [review]

File: src/server/scripts/Kalimdor/RuinsOfAhnQiraj/boss_ossirian.cpp (hunk 1 above)
Comment (UltraNix) [reply src/server/scripts/Kalimdor/RuinsOfAhnQiraj/boss_ossirian.cpp:110]: No. It's not comparison, but assigment.

**Comment by UltraNix** [comment]
@acidmanifesto Should be okay.
@Gultask Please, test it again.

**Comment by Gultask** [comment]
The starting crystal is respawning too fast, otherwise that one in particular is ok. It should respawn at the same time as the boss after the hard reset, 20-30s later.

When pulling the boss, the same crystal spawns every time and it's not a random one.

If I use the crystal before the boss reaches its range, it will not spawn another crystal.

https://user-images.githubusercontent.com/100873791/184508240-3f76b7ff-f85d-4d88-8a33-56d2e5e206b1.mp4

**Comment by UltraNix** [comment]
@Gultask 
https://wowpedia.fandom.com/wiki/Ossirian_the_Unscarred
```
Generally crystals spawn in a clockwise or counterclockwise manner, and the kite path can be pretty consistent. Some crystals spawn pretty close to each other, so the tank should just get to the next crystal, stay there until the 15 second warning and let the rest of the raid catch up and do damage to Ossirian
```
So it can't be random.

**Comment by UltraNix** [comment]
@Gultask 
`When pulling the boss, the same crystal spawns every time and it's not a random one.`
Where did you get this one?

https://wowpedia.fandom.com/wiki/Ossirian_the_Unscarred
`Generally crystals spawn in a clockwise or counterclockwise manner, and the kite path can be pretty consistent. Some crystals spawn pretty close to each other, so the tank should just get to the next crystal, stay there until the 15 second warning and let the rest of the raid catch up and do damage to Ossirian`

**Comment by UltraNix** [comment]
@Gultask Try now.

**Comment by Maelthyrr** [comment]
Tested:

- [X] Random spawn crystal
- [X If I use the crystal before the boss reaches its range, it's respawn another crystal
- [X] Cystal work correctly: remove boss buff and apply a vulnerability
- [X] Reset work

**Review by Nyeriah — APPROVED** [review]

**Review by acidmanifesto — APPROVED** [review]

**Pull request merged**